elfpy/markets/hyperdrive/hyperdrive_market.py

```diff
@@ -642,3 +640,169 @@ def apply_checkpoint(self, checkpoint_time: float, share_price: float) -> float:
+            If true, the user is paid in underlying, if false the contract transfers in yield source
+            directly. Note - for some paths one choice may be disabled or blocked.
+
+        Returns
+        -------
+        float
+            The amount of base the LP received.
+        """
+
+        # create a new checkpoint if necessary, close positions at the checkpoint time one
+        # position_duration ago.
+        self.apply_checkpoint(self.latest_checkpoint_time, self.market_state.share_price)
+        market_deltas, wallet_deltas = self.calc_redeem_withdraw_shares(shares, min_output, as_underlying)
+        self.update_market(market_deltas)
+        agent_wallet.update(wallet_deltas)
+
+        return wallet_deltas.balance.amount
+
+    def calc_redeem_withdraw_shares(
+        self, shares: float, min_output: float, as_underlying: bool
+    ) -> tuple[hyperdrive_actions.MarketDeltas, wallet.Wallet]:
+        r"""
+        Calculates the market and wallet deltas for redeemable withdrawal shares, if enough margin
+        has been freed to do so.
+
+        Parameters
+        ----------
+        shares: float
+            The withdrawal shares to redeem.
+        min_output: float
+            The minimum amount of base the LP expects to receive.
+        as_underlying: bool
+            If true, the user is paid in underlying, if false the contract transfers in yield source
+            directly. Note - for some paths one choice may be disabled or blocked.
+
+        Returns
+        -------
+        tuple[hyperdrive_actions.MarketDeltas, wallet.Wallet]
+
+        """
+        market_deltas = hyperdrive_actions.MarketDeltas()
+        wallet_deltas = wallet.Wallet(address=0)
+
+        # We burn the shares from the user
+        wallet_deltas.withdraw_shares -= shares
+
+        # The user gets a refund on their margin equal to the face value of their withdraw shares
+        # times the percent of the withdraw pool which has been lost.
+        recovered_margin = (
+            shares * self.market_state.withdraw_capital / self.market_state.withdraw_shares_ready_to_withdraw
+        )
+
+        # The user gets interest equal to their percent of the withdraw pool times the withdraw pool
+        # interest
+        recovered_interest = (
+            shares * self.market_state.withdraw_interest / self.market_state.withdraw_shares_ready_to_withdraw
+        )
+
+        # Update the pool state
+        # Note - Will revert here if not enough margin has been reclaimed by checkpoints or by
+        #  position closes
+        market_deltas.withdraw_shares_ready_to_withdraw -= shares
+        market_deltas.withdraw_capital -= recovered_margin
+        market_deltas.withdraw_interest -= recovered_interest
+
+        # Withdraw for the user
+        base_proceeds = self._withdraw(recovered_margin + recovered_interest, as_underlying)
+        # TODO: figure out how to keep track of hyperdrive's base asset amount.  market_deltas has
```

Comment:
Can't you add `base_reserves` to `hyperdrive_market.MarketState`?

Reply:
I assume the reason why you can't compute `base_reserves = share_reserves * share_price` is because the share price is variable and so you'd want to do track update the base reserves each time withdraw_shares are redeemed?

Reply:
yeah it's not as simple as that. share_reserves are in shares, not base.  I think the only base would come from the withdraw pool.  there are spots in the solidity where they use the base token balance of the contract, so we'll need the same thing but gotta figure out where that balance comes from first.